### PR TITLE
Include program comments in program parsetree

### PR DESCRIPTION
**Pull request by ospencer**
This PR adds program comments to the parsetree. This will allow build tools that look at parsed programs to also read comments, for things like documentation or formatting.

### CODE REVIEW

**Comment by phated** [comment]
While writing this, we ran into bugs when we made changes to `String_utils.slice`, so I'd like to try to set up inline testing for our utility methods before we call this ready

File: compiler/test/test_end_to_end.re [hunk 1]
```diff
@@ -1283,47 +1284,83 @@ let string_tests = {
     tparse(
       "string_parse_dqs1",
       "\"foo\"",
-      {statements: [str("foo")], prog_loc: Location.dummy_loc},
+      {
```
Comment (belph) [inline compiler/test/test_end_to_end.re:1287]: @ospencer Maybe add a test case?

**Review by belph — CHANGES_REQUESTED** [review]
Great stuff! LGTM; just one small request. Also, we should think about block comments soon.

**Comment by ospencer** [comment]
@belph I dunno if you remember, but the plan is to just swap our comment syntax to C-style.

**Comment by ospencer** [comment]
Made https://github.com/grain-lang/grain/issues/275 to track.

File: compiler/test/test_end_to_end.re (hunk 1 above)
Comment (phated) [reply compiler/test/test_end_to_end.re:1287]: Good call here. I am trying to write these now and it seems like a bunch of comments are being added to this block.

Investigating.

Comment (phated) [reply compiler/test/test_end_to_end.re:1295]: I added a test for this in a separate block (see individual commit).

**Comment by phated** [comment]
@ospencer @belph Alright, I believe I've added enough tests and fixed an issue I came across now. I'll re-request reviews from you both, so please review when you have a chance.

**Pull request merged**